easybuild/easyblocks/generic/gopackage.py

```diff
@@ -0,0 +1,169 @@
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for Go packages, implemented as an EasyBlock
+
+@author: Pavel Grochal (INUITS)
+"""
+import os
+from distutils.version import LooseVersion
+
+import easybuild.tools.environment as env
+from easybuild.framework.easyblock import EasyBlock
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.tools.filetools import which
+from easybuild.tools.modules import get_software_root, get_software_version
+from easybuild.tools.run import run_cmd
+
+
+class GoPackage(EasyBlock):
+    """Builds and installs a Go package, and provides a dedicated module file."""
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Easyconfig parameters specific to Go packages."""
+        if extra_vars is None:
+            extra_vars = {}
+        extra_vars.update({
+            'modulename': [None, "Module name of the go package, when building non-native module", CUSTOM],
+            'forced_deps': [None, "Force specific version of go package, when building non-native module", CUSTOM],
+        })
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        """Initialize custom class variables."""
+        super(GoPackage, self).__init__(*args, **kwargs)
+        self.go_cmd = None
+
+    def prepare_go(self):
+        """Go-specific preperations."""
+
+        go = None
+        go_root = get_software_root('Go')
+        if go_root:
+            bin_go = os.path.join(go_root, 'bin', 'go')
+            if os.path.exists(bin_go) and os.path.samefile(which('go'), bin_go):
+                # if Go is listed as a (build) dependency, use 'go' command provided that way
+                go = os.path.join(go_root, 'bin', 'go')
+                self.log.debug("Retaining 'go' command for Go dependency: %s", go)
+        if go:
+            self.go_cmd = go
+        else:
+            raise EasyBuildError("Failed to pick go command to use. Is it listed in dependencies?")
+
+        go_version = get_software_version('Go')
+        if LooseVersion(go_version) < LooseVersion("1.11"):
+            raise EasyBuildError("Go version < 1.11 doesn't support installing modules from go.mod")
+
+        # enforce use of go modules
+        env.setvar('GO111MODULE', 'on', verbose=False)
+        # set bin folder
+        env.setvar('GOBIN', os.path.join(self.installdir, 'bin'), verbose=False)
+
+    def configure_step(self):
+        """Configure Go package build/install."""
+
+        if self.go_cmd is None:
+            self.prepare_go()
```

Comment:
You can drop this if you use `prepare_step` above, and just assume `self.go_cmd` is set by now